### PR TITLE
Config: unified logic to retrieve text config

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Expected in #32673, #33212

This PR unifies how we access the text config in composite models in a single function, instead of having to look for the right key. This avoids recurrent patterns of `foo if hasattr(...) else bar`.

[Previously discussed with @amyeroberts [here](https://github.com/huggingface/transformers/pull/32673#discussion_r1719792999)]

### CODE REVIEW

File: utils/tests_fetcher.py [hunk 1]
```diff
@@ -1054,7 +1054,7 @@ def infer_tests_to_run(
         # Then we grab the corresponding test files.
         test_map = create_module_to_test_map(reverse_map=reverse_map, filter_models=filter_models)
         for f in modified_files + impacted_files:
-            if f in test_map:
+            if f in test_map and test_map[f] is not None:
```
Comment (gante) [inline utils/tests_fetcher.py:1057]: `test_map[f]` seems to be empty in some situations, which breaks the line below.

More specifically, in the commit before this change (`f4e04b7f519cf8c79b80233cd7dbdd17e9586038`), this was the CI for the test fetcher: https://app.circleci.com/pipelines/github/huggingface/transformers/102539/workflows/ebaf3657-9fab-4718-8240-ad1f350eb9cf/jobs/1365693

File: tests/test_modeling_common.py [hunk 2]
```diff
@@ -1747,12 +1747,13 @@ def test_resize_position_vector_embeddings(self):
             self.assertTrue(models_equal)
 
     def test_resize_tokens_embeddings(self):
+        if not self.test_resize_embeddings:
+            self.skipTest(reason="test_resize_embeddings is set to `False`")
```
Comment (gante) [inline tests/test_modeling_common.py:1751]: (moved the skip here: no point in spending compute if we are going to skip the test)

File: tests/test_pipeline_mixin.py [hunk 3]
```diff
@@ -675,14 +675,12 @@ def validate_test_components(test_case, task, model, tokenizer, processor):
     # Avoid `IndexError` in embedding layers
     CONFIG_WITHOUT_VOCAB_SIZE = ["CanineConfig"]
     if tokenizer is not None:
-        config_vocab_size = getattr(model.config, "vocab_size", None)
+        # Removing `decoder=True` in `get_text_config` can lead to conflicting values e.g. in MusicGen
```
Comment (gante) [inline tests/test_pipeline_mixin.py:678]: Alternatively, we can add a flag to a) return the first valid match; OR b) return all matches when there is more than one match.

File: src/transformers/configuration_utils.py [hunk 4]
```diff
@@ -1057,6 +1056,36 @@ def _get_non_default_generation_parameters(self) -> Dict[str, Any]:
 
         return non_default_generation_parameters
 
+    def get_text_config(self, decoder=False) -> "PretrainedConfig":
+        """
+        Returns the config that is meant to be used with text IO. On most models, it is the original config instance
+        itself. On specific composite models, it is under a set of valid names.
+
+        If `decoder` is set to `True`, then only search for decoder config names.
```
Comment (zucchini-nlp) [inline src/transformers/configuration_utils.py:1064]: for my own understanding, what does it mean to search in "decoder config names"? Is it somehow related to a model being an encoder-decoder or decoder-only?

From what I see, `text_encoder` is used in Musicgen only and we never used the `decoder=False` in transformers

**Review by zucchini-nlp — APPROVED** [review]
Super cool, thanks for making the code cleaner! One tiny question about the `decoder` arg, it is not clear to me why it's needed?

File: src/transformers/configuration_utils.py (hunk 4 above)
Comment (gante) [reply src/transformers/configuration_utils.py:1064]: It is indeed mostly for `musicgen` at the moment. Needing/wanting to use the encoder text config is uncommon, but there is at least one pair of use cases:
1. `resize_token_embeddings` (has `decoder=False`)
2. The tests for `resize_token_embeddings`

File: src/transformers/configuration_utils.py [hunk 5]
```diff
@@ -1026,10 +1026,9 @@ def _get_non_default_generation_parameters(self) -> Dict[str, Any]:
         try:
             default_config = self.__class__()
         except ValueError:
-            for decoder_attribute_name in ("decoder", "generator", "text_config"):
-                if hasattr(self, decoder_attribute_name):
-                    default_config = getattr(self, decoder_attribute_name).__class__()
-                    break
+            decoder_config = self.get_text_config(decoder=True)
+            if decoder_config is not self:
+                default_config = decoder_config.__class__()
```
Comment (LysandreJik) [inline src/transformers/configuration_utils.py:1031]: I think the comment above would benefit from an update to clarify what's happening here

File: src/transformers/configuration_utils.py [hunk 6]
```diff
@@ -1057,6 +1056,36 @@ def _get_non_default_generation_parameters(self) -> Dict[str, Any]:
...
+        for text_config_name in possible_text_config_names:
+            if hasattr(self, text_config_name):
+                text_config = getattr(self, text_config_name, None)
+                if text_config is not None:
+                    valid_text_config_names += [text_config_name]
+
+        if len(valid_text_config_names) > 1:
+            raise ValueError(
+                f"Multiple valid text configs were found in the model config: {valid_text_config_names}. "
+                "Either don't use `get_text_config`, as it is ambiguous -- access the text configs directly."
```
Comment (LysandreJik) [inline src/transformers/configuration_utils.py:1083]: I don't entirely understand what this comment hints at doing :sweat_smile:

**Review by LysandreJik — APPROVED** [review]
Thanks for the PR @gante

File: src/transformers/configuration_utils.py (hunk 6 above)
Comment (gante) [reply src/transformers/configuration_utils.py:1083]: Indeed, I've updated the code but left this exception halfway 😅

**Pull request merged**